contributors/design-proposals/auditing.md

```diff
@@ -0,0 +1,361 @@
+    AuditHttpHeader AuditLevel   = "HttpHeader"
+    // AuditRequestBody provides Header level of auditing, and additionally
+    // logs unstructured request body
+    AuditRequestBody AuditLevel = "RequestBody"
+    // AuditResponseBody provides Request level of auditing, and additionally
+    // logs unstructured response body
+    AuditResponseBody AuditLevel = "ResponseBody"
+    // AuditStorageobject provides Response level, and additionally
+    // logs object before and after saving in storage
+    AuditStorageObject AuditLevel = "StorageObject"
+)
+
+type Event struct {
+    // AuditLevel at which event was generated
+    Level AuditLevel
+
+    // below fields are filled at HttpHeader level and higher:
+
+    // Unique audit ID, generated for each request
+    ID string
+    // Time the event reached the apiserver
+    Timestamp Timestamp
+    // Source IP, from where the request originates
+    SourceIP string
+    // HTTP method sent by the client
+    HttpMethod string
+    // Authentication method used to allow users access the cluster
+    AuthMethod string
+    // RequestURI is the Request-Line as sent by the client to a server
+    RequestURI string
+    // User information
+    User UserInfo
+    // Impersonate-User header value
+    AsUser string
+    // Impersonate-Group header value
+    AsGroups string
+    // Object reference this request is targeted at
+    Object ObjectReference
+    // Response status code are returned by the server
+    ResponseStatusCode int
+
+    // below fields are filled at RequestObject level and higher:
+
+    // RequestObject logged before admission
+    RequestObject runtime.Unstructured
+
+    // below fields are filled at StorageObject level and higher:
+
+    // Object value before modification (will be empty when creating new object)
+    OldObject runtime.Object
+    // Object value after modification (will be empty when removing object)
+    NewObject runtime.Object
+    // Difference between old and new object
+    Patch     []byte
+}
+```
+
+### Concurrency/Transaction Model for Events
+
+TODO: define (and maybe adapt interface) how synchronization of the writable audit event of a context takes place.
+
+### Policy
+
+The audit policy determines what audit event is generated for a given request. The policy is
+configured by the cluster administrator. Here is a sketch of the policy API:
+
+```go
+type Policy struct {
+    // Rules specify the audit Level a request should be recorded at.
+    // A request may match multiple rules, in which case the FIRST matching rule is used.
+    // The default audit level is None, but can be overridden by a catch-all rule at the end of the list.
+    Rules []PolicyRule
+
+    // Discussed under Filters section.
+    Filters []Filter
+}
+
+// Based off the RBAC PolicyRule
+type PolicyRule struct {
+    // Required. The Level that requests matching this rule are recorded at.
+    Level Level
```

Comment:
If we go this root, then I think a more structured approach would make more sense, e.g.:
```go
type Action struct {
  Metadata bool  // Resource, user, & request identifiers (a.k.a. "HttpHeader")
  RequestBody bool
  ResponseBody bool
  StorageObject bool
}
```

My main concern with this is that it makes the configuration much heavier. It might make sense for the internal datatype, but as a user is there really a usecase for logging the response but not the metadata?

Another consideration is that once `Filter`s are implemented, they can provide that further customization while keeping the basic usecase lightweight. E.g. specify `ResponseBody` level, and then add a Filter to drop the Request.

I suppose `StorageObject` is a bit orthogonal to Request & Response, but TBH I'm not sure I understand the usecase for providing the StorageObject. When would the Response object be insufficient?